### PR TITLE
feat(tls): implement non-blocking TrySend with async flush offload

### PR DESCRIPTION
**Pull request by glevkovich**
Implemented TlsSocket::TrySend using a non-blocking state machine loop to handle upstream backpressure and TLS state requirements (NEED_READ/NEED_WRITE) without unnecessary context switching.

Key changes include:
- Async Flush Offload: Introduced a "Fire and Hold" mechanism. If TrySend successfully consumes all user data but fails to fully flush the pending ciphertext to the network (partial write), the remaining flush is offloaded to a detached background AsyncReq. This prevents stalling the caller while ensuring data safety.

- Async Logic Update: Updated AsyncRoleBasedAction to correctly handle detached flush requests (where vec == nullptr), allowing the background fiber to exit gracefully once the buffer is drained.

- Small Buffer Optimization (SSO): Applied SBO using absl::InlinedVector for iovec copying to minimize heap allocations for standard batch sizes.

- Build System: Added iovec_utils.cc to the tls_lib target in CMakeLists.txt.

Testing improvements:
- Scatter-Gather Tests: Added TrySendVectorTest to validate behavior with various iovec counts and split patterns.

- White-Box Testing: Introduced MockTlsSocketTest infrastructure with Strict/Nice mocks for the Proactor, FiberSocket, and TlsEngine.

- Edge Case Coverage: Added TrySendErrorTest to verify handling of EAGAIN, Dirty Shutdowns, and Concurrency Conflicts.

- Async Verification: Added TrySendAsyncFlushTest to verify that stranded data is correctly offloaded to the async path.

- Bug Fix: Fixed RegisterOnRecv test to explicitly call ResetOnRecvHook() before manual TryRecv, resolving a "Concurrent TryRecv and Recv" usage error.

### CODE REVIEW

File: util/tls/tls_socket.cc [hunk 1]
```diff
@@ -505,13 +505,157 @@ void TlsSocket::OnRecv(const RecvNotification& rn, const OnRecvCb& recv_cb) {
...
+    // 2. Push data from user buffer into the engine
+    DCHECK_EQ(engine_->OutputPending(), 0u);
+    DCHECK_GT(iov_cursor->iov_len, 0u);
+    PushResult push_result{PushToEngine(iov_cursor, curr_iovec_len)};
+    // PushToEngine Result Semantics:
+    // 1. written > 0:   Bytes successfully consumed from the user buffer. This happens even if an
+    // error/requirement (opcode < 0) immediately follows.
+    // 2. opcode < 0:    Engine requires action (NEED_READ/WRITE) or failed (EOF).
+    // If written > 0 AND opcode < 0, it means a partial write occurred before the stop.
+    // 3. opcode == 0:   Success. All bytes in this chunk were consumed.In this case, we expect
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket.cc]: fixed

File: util/tls/tls_socket_test.cc [hunk 2]
```diff
@@ -1126,5 +1248,260 @@ INSTANTIATE_TEST_SUITE_P(
...
+        if (*p.sock_recv_ret && (**p.sock_recv_ret > 0)) {
+          EXPECT_CALL(*mock_engine_, CommitInput(**p.sock_recv_ret)).Times(1);
+        }
+      }
+    }
+
+    // Execution
+    uint8_t user_data[] = "payload";
+    iovec v = {user_data, sizeof(user_data)};
+    EXPECT_FALSE(sock_->IsOpen()) << "Mock socket should assume open state via mocks";
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket_test.cc]: fixed

File: util/tls/tls_socket.cc [hunk 3]
```diff
@@ -505,13 +505,173 @@ void TlsSocket::OnRecv(const RecvNotification& rn, const OnRecvCb& recv_cb) {
...
+    return 0;  // nothing to send
+  bool has_data{false};
+  for (size_t i{}; i < len; ++i) {
+    if (v[i].iov_len > 0) {
+      has_data = true;
+      break;
+    }
+  }
+  if (!has_data)
+    return 0;  // nothing to send
```
Comment (romange) [inline util/tls/tls_socket.cc]: can be factored out into a helper function, and lets DCHECK as well, as I do not see why should we allow callers to call this function with no data

File: util/tls/tls_socket_test.cc [hunk 4]
```diff
@@ -315,6 +327,118 @@ TEST_P(TlsSocketTest, RegisterOnRecv) {
...
+// edge cases such as partial sends, chunking, and large iovec counts (including 1-byte chunks).
+TEST_P(TrySendVectorTest, SendScatterGather) {
+  // 1. Parse Parameters: "engine:count" (e.g., "epoll:3")
+  string_view param{GetParam()};
+  size_t sep_pos{param.find(':')};
+  ASSERT_NE(sep_pos, string_view::npos)
+      << "Invalid test parameter '" << param << "'. Expected format 'engine:count'";
+  size_t target_iovec_count{};
+  auto count_str{param.substr(sep_pos + 1)};
+  auto [ptr, ec] = std::from_chars(count_str.data(), count_str.data() + count_str.size(), target_iovec_count);
```
Comment (romange) [inline util/tls/tls_socket_test.cc]: we use `absl::SimpleAtoi` for that

File: util/tls/tls_socket.cc [hunk 5]
```diff
@@ -505,13 +505,173 @@ void TlsSocket::OnRecv(const RecvNotification& rn, const OnRecvCb& recv_cb) {
...
+    // Another fiber is currently writing, we cannot safely proceed
+    DVSOCK(3) << "TrySend blocked: WRITE_IN_PROGRESS detected";
+    return make_unexpected(make_error_code(errc::resource_unavailable_try_again));
+  }
+  bool read_in_progress{(state_ & READ_IN_PROGRESS) != 0};
+  size_t total_bytes_sent{};
+  std::error_code returned_status{};
+  constexpr size_t kStackIovecs = 16;
+  iovec stack_vec[kStackIovecs];
+  std::vector<iovec> heap_vec;  // Fallback container if the user sends a huge iovec array
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket.cc]: thanks, will change.

File: util/tls/tls_socket_test.cc (hunk 4 above)
Comment (glevkovich) [reply util/tls/tls_socket_test.cc]: changed to SimpleAtoi

File: util/tls/tls_socket_test.cc [hunk 6]
```diff
@@ -315,6 +327,118 @@ TEST_P(TlsSocketTest, RegisterOnRecv) {
...
+  ASSERT_EQ(ec, std::errc{}) << "Failed to parse iovec count from '" << count_str << "'";
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket_test.cc]: fixed.

File: util/tls/tls_socket.cc [hunk 7]
```diff
@@ -505,13 +505,173 @@ void TlsSocket::OnRecv(const RecvNotification& rn, const OnRecvCb& recv_cb) {
...
+  }
+  if (!has_data)
+    return 0;  // nothing to send
+  bool write_in_progress{(state_ & WRITE_IN_PROGRESS) != 0};
+  if (write_in_progress) {
+    // Another fiber is currently writing, we cannot safely proceed
+    DVSOCK(3) << "TrySend blocked: WRITE_IN_PROGRESS detected";
+    return make_unexpected(make_error_code(errc::resource_unavailable_try_again));
+  }
+  bool read_in_progress{(state_ & READ_IN_PROGRESS) != 0};
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket.cc]: I don't think so:

- TrySend does not control SSL engine exclusively. I have no idea who called the engine before, maybe via other function. Maybe oms read on the socket generated SSL metadata on the output , and other cases.
- TrySend can run concurrently with some *Read* operation, so again OutputPending() can become non-zero.

File: util/tls/tls_socket.cc [hunk 8]
```diff
@@ -505,13 +505,173 @@ void TlsSocket::OnRecv(const RecvNotification& rn, const OnRecvCb& recv_cb) {
...
+  // We make a local mutable copy of the iovec descriptors because AdvanceIovec
+  // modifies them (adjusting base pointers and lengths) to track partial writes.
+  // The input array 'v' is const and belongs to the caller, so we cannot modify it.
+  std::memcpy(curr_iov, v, len * sizeof(iovec));
+  iovec* iov_cursor = curr_iov;
+  uint32_t curr_iovec_len{len};
+
+  while ((curr_iovec_len > 0) || (engine_->OutputPending() > 0)) {
+    // 1. Flush into the upstream socket any pending output from the engine output buffer before
+    // pushing more data to the engine from the user. These might be bytes from previous call.
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket.cc]: 1. Regarding Batching: You are right that the current "Flush-Push-Flush" loop could result in small packets (one per iovec), which isn't ideal for performance.  I will try to refactor this to Send only once in multiple PushToEngine calls before triggering a flush to the upstream socket. If this makes the non-blocking logic too complex for this single PR, I will merge this correct version first and optimize this in a dedicated follow-up PR to keep the changes manageable.

2. Regarding OutputPending & Async Flushing: I have to disagree on the "asynchronous process" point. At the very top of TrySend, I check:
`if ((state_ & WRITE_IN_PROGRESS) != 0) return ... try_again;`
If there were any asynchronous process or background fiber currently flushing this buffer, WRITE_IN_PROGRESS would be set, and I would have exited immediately.
Since I reached this line, I am the exclusive writer. There is no other active process responsible for this data. If I don't flush OutputPending here, the data will sit in the SSL BIO indefinitely (causing latency) until the next API call happens to trigger a write. Therefore, it is my responsibility to flush it. Also, this is used as a single place to flush between iteration (only when needed, in the optimised-to-be-written version).

File: util/tls/tls_socket.cc [hunk 9]
```diff
@@ -505,13 +507,138 @@ void TlsSocket::OnRecv(const RecvNotification& rn, const OnRecvCb& recv_cb) {
...
+      auto output_buf{engine_->PeekOutputBuf()};
+      DCHECK(!output_buf.empty());
+      auto send_result{next_sock_->TrySend(output_buf)};
+      if (send_result) {
+        CHECK_LE(*send_result, output_buf.size());
+        engine_->ConsumeOutputBuf(*send_result);
+        DVSOCK(3) << "Flushed " << *send_result << " bytes to upstream";
+        if ((*send_result) < output_buf.size()) {  // case 1.A: partial write
+          // upstream socket is full - try again later
+          returned_status = make_error_code(errc::resource_unavailable_try_again);
```
Comment (romange) [inline util/tls/tls_socket.cc]: 1. Bug. what ensures that the write will be flushed when `next_sock_` becomes available?
2. `make_error_code(errc::resource_unavailable_try_again);` is wrong as you already consumed some of the input data and it was copied to ssl engine. From a caller perspective - they need to retry the entire operation, so they will try to push the same data again

File: util/tls/tls_socket.cc [hunk 10]
```diff
@@ -505,13 +507,138 @@ void TlsSocket::OnRecv(const RecvNotification& rn, const OnRecvCb& recv_cb) {
...
+  vec[0].iov_base = const_cast<uint8_t*>(buf.data());
+  vec[0].iov_len = buf.size();
+  return TrySend(vec, 1);
 }
 
 io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
-  LOG(DFATAL) << "Not implemented";
-  return 0;
+  if (IsEmptyIovec(v, len)) {
+    DCHECK(false) << "TrySend with empty iovec";
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket.cc]: fixed

File: util/tls/tls_socket_test.cc [hunk 11]
```diff
@@ -315,6 +328,118 @@ TEST_P(TlsSocketTest, RegisterOnRecv) {
...
+  // Examples:
+  // If target is 3:      16384 / 3     = ~5461 bytes per chunk
+  // If target is 16384:  16384 / 16384 = 1 byte per chunk
+  size_t kChunkSize = (kPayload.size() + target_iovec_count - 1) / target_iovec_count;
+  if (kChunkSize == 0)
+    kChunkSize = 1;
+  std::vector<iovec> vecs;
+  vecs.reserve((kPayload.size() + kChunkSize - 1) / kChunkSize);
+  char* src = kPayload.data();
+  size_t remaining = kPayload.size();
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket_test.cc]: that was probably a name typo, renaming.

File: util/tls/tls_socket_test.cc [hunk 12]
```diff
@@ -4,12 +4,14 @@
 
 #include "util/tls/tls_socket.h"
 
+#include <absl/strings/numbers.h>
+#include <absl/strings/str_cat.h>
 #include <gmock/gmock.h>
 
 #include <algorithm>
+#include <charconv>
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket_test.cc]: fixed

File: util/tls/tls_socket_test.cc [hunk 13]
```diff
@@ -315,6 +328,118 @@ TEST_P(TlsSocketTest, RegisterOnRecv) {
...
+          if (written >= current_iov->iov_len) {  // consumed all of current iov
+            written -= current_iov->iov_len;
+            current_iov++;
+            current_count--;
+          } else {  // partially consumed current iov
+            current_iov->iov_base = static_cast<char*>(current_iov->iov_base) + written;
+            current_iov->iov_len -= written;
+            written = 0;
+          }
+        }
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket_test.cc]: fixed

File: util/tls/tls_socket_test.cc [hunk 14]
```diff
@@ -294,13 +299,27 @@ TEST_P(TlsSocketTest, RegisterOnRecv) {
...
+        break;
+      }
+    }
   });
 
   // Launch a fiber to perform the actual TryRecv and store the result.
   auto recv_fb = proactor_->LaunchFiber([&] {
     data_ready.Wait();
+
+    // FIX: Unregister the callback before invoking TryRecv
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket_test.cc]: That's a leftover (work in progress)  from previous  wrong implementation, the current comment is wrong and irrelevant,but the de-register is correct.
Since we used RegisterOnRecv few lines above (which is a "push" callback mode) , calling TryRecv while still registered is illegal and may cause the fatal error in tls_socket.cc:
`LOG(DFATAL) << "Concurrent TryRecv and Recv detected - this is a usage error.";`

If I want to  use TryRecv I must de-register 1st. I'll fix the comment:
```
// Unregister the callback before invoking TryRecv.
// The underlying socket cannot handle manual 'TryRecv' calls while 
// a 'RegisterOnRecv' hook is active (Push vs Pull mode conflict).
```

File: util/tls/tls_socket_test.cc [hunk 15]
```diff
@@ -1126,5 +1245,418 @@ INSTANTIATE_TEST_SUITE_P(
         }
 #endif
         ));
+struct TrySendScenario {
```
[the remark this answers is not in the document]
Comment (glevkovich) [reply util/tls/tls_socket_test.cc]: I think we are both 'right' here. I know about the maintenance fee.

These tests are tightly coupled to the implementation details, which definitely adds a maintenance cost. In my past work, this was usually a fee I am willing to pay to ensure high-quality unit testing. I've found that relying solely on integration tests makes it nearly impossible to deterministically reproduce specific edge cases - such as EAGAIN timing, partial writes, or dirty shutdowns.

I believe this is a necessary tradeoff: we pay a higher maintenance 'fee' on these tests to ensure we do not regress on subtle behaviors that are otherwise un-testable. I would prefer to keep them to 'guard' against breaking these complex error paths, but I am open to suggestions if you see a cleaner way to verify these scenarios.

If you feel we should not take on this maintenance burden, the only alternative would be to remove most /all of the mock tests. I fully understand that decision. Still,  I believe there is a good chance we will pay that price later (implicitly) by spending more time debugging TLS during or after changes because we broke an expected behavior that the mock tests would have caught.

**Review by romange — APPROVED** [review]
